blockchain/storage/testing.go

```diff
@@ -469,31 +469,36 @@ type writeReceiptsDelegate func(types.Hash, []*types.Receipt) error
 type readReceiptsDelegate func(types.Hash) ([]*types.Receipt, error)
 type writeTxLookupDelegate func(types.Hash, types.Hash) error
 type readTxLookupDelegate func(types.Hash) (types.Hash, bool)
+type writeLastProcessedEvent func(string, string) error
+type readLastProcessedEvent func(string) (string, bool)
 type closeDelegate func() error
 
 type MockStorage struct {
-	readCanonicalHashFn    readCanonicalHashDelegate
-	writeCanonicalHashFn   writeCanonicalHashDelegate
-	readHeadHashFn         readHeadHashDelegate
-	readHeadNumberFn       readHeadNumberDelegate
-	writeHeadHashFn        writeHeadHashDelegate
-	writeHeadNumberFn      writeHeadNumberDelegate
-	writeForksFn           writeForksDelegate
-	readForksFn            readForksDelegate
-	writeTotalDifficultyFn writeTotalDifficultyDelegate
-	readTotalDifficultyFn  readTotalDifficultyDelegate
-	writeHeaderFn          writeHeaderDelegate
-	readHeaderFn           readHeaderDelegate
-	writeCanonicalHeaderFn writeCanonicalHeaderDelegate
-	writeBodyFn            writeBodyDelegate
-	readBodyFn             readBodyDelegate
-	writeSnapshotFn        writeSnapshotDelegate
-	readSnapshotFn         readSnapshotDelegate
-	writeReceiptsFn        writeReceiptsDelegate
-	readReceiptsFn         readReceiptsDelegate
-	writeTxLookupFn        writeTxLookupDelegate
-	readTxLookupFn         readTxLookupDelegate
-	closeFn                closeDelegate
+	readCanonicalHashFn       readCanonicalHashDelegate
+	writeCanonicalHashFn      writeCanonicalHashDelegate
+	readHeadHashFn            readHeadHashDelegate
+	readHeadNumberFn          readHeadNumberDelegate
+	writeHeadHashFn           writeHeadHashDelegate
+	writeHeadNumberFn         writeHeadNumberDelegate
+	writeForksFn              writeForksDelegate
+	readForksFn               readForksDelegate
+	writeTotalDifficultyFn    writeTotalDifficultyDelegate
+	readTotalDifficultyFn     readTotalDifficultyDelegate
+	writeHeaderFn             writeHeaderDelegate
+	readHeaderFn              readHeaderDelegate
+	writeCanonicalHeaderFn    writeCanonicalHeaderDelegate
+	writeBodyFn               writeBodyDelegate
+	readBodyFn                readBodyDelegate
+	writeSnapshotFn           writeSnapshotDelegate
+	readSnapshotFn            readSnapshotDelegate
+	writeReceiptsFn           writeReceiptsDelegate
+	readReceiptsFn            readReceiptsDelegate
+	writeTxLookupFn           writeTxLookupDelegate
+	readTxLookupFn            readTxLookupDelegate
+	writeLastProcessedEventFn writeLastProcessedEvent
+	readLastProcessedEventFn  readLastProcessedEvent
```

Comment:
Now these two mock methods seem to be called from the tests in `rootchain/sam` package. I think it doesn't make sense to define them in MockStorage. You can define another MockStorage inside `rootchain/sam`.

Reply:
There is actually a separate mock storage implementation for the SAMUEL storage:
https://github.com/0xPolygon/polygon-edge/blob/95c8a400ba84901a0f867498e369b620de161f50/rootchain/samuel/mock_test.go#L150-L172

What you linked is required because `MockStorage` needs to implement the base `Storage` interface methods to be used as a `Storage` interface

Reply:
Firstly, the added methods and the other functions are called from different modules completely. Why not consider about splitting interface?

Reply:
The initialization is not present in this PR, but in #713, the DB interface is going to be of type `Storage` (key-value store), so we need to keep it